flashcards/backend/decks/models.py

```diff
@@ -20,3 +20,42 @@ def __str__(self):
     class Meta:
         verbose_name = "category"
         verbose_name_plural = "categories"
+
+
+class Tag(models.Model):
+    """Model for representing tags for decks
+    :param name: Tag name
+    :type name: str
+    :param description: Description for tag
+    :type description: str
+    :param deck: Foreign Key for Deck model
+    :type deck: int
+    :param flashcard: Foreign Key for Flashcard model
+    :type flashcard: int
+    """
+    name = models.CharField(max_length=24)
+    description = models.TextField()
+    deck = models.ForeignKey('Deck', on_delete=models.DO_NOTHING)
+    flashcard = models.ForeignKey('Flashcard', on_delete=models.DO_NOTHING)
+
+    def __str__(self):
+        return self.name
+
+    def save(self, *args, **kwargs):
+        tag = Tag.objects.filter(name=self.name.lower())
+        self.name = self.name.lower()
+
+        if not tag:
+            return super().save(*args, **kwargs)
+
+        return tag
```

Comment:
Check this out

```python
from django.db.models.functions import Lower

constraints = [
            models.UniqueConstraint(Lower('name'), name='unique_lower_name_tag')
        ]
```

Reply:
But it blocks the record only from the CMS level, still allows you to write to the database from the hand of different sizes of letters in the same names.

Reply:
Because to trigger the validator you have to call the `full_clean` method on save or in test.  So I think what you can do is the following:

```python
def save(self, *args, **kwargs):
    self.name = self.name.lower()
    self.full_clean()
return super().save(*args, **kwargs)
```

or in test:

```python
def test_lower_upper_uniqness():  # just an egxample name
    # create test case with creating duplicated name objects
    object_name.full_clean()  # this should trigger the validators
```

As we've learnd recently validators aren't triggered automatically on simple object save to DB.

Reply:
Ok, but it doesn't seem to be needed in this case. The custom save method can't be cheated ;). But I added, it's a good piece of code and might come in handy later.